src/ws/cockpitwebservice.c

```diff
@@ -598,14 +611,60 @@ process_transport_init (CockpitWebService *self,
         json_object_unref (self->init_received);
       self->init_received = json_object_ref (options);
 
+      if (cockpit_json_get_object (options, "capabilities", NULL, &capabilities) && capabilities)
+        {
+          if (!cockpit_json_get_bool (capabilities, "explicit-superuser", FALSE, &explicit_superuser_capability))
+            g_warning ("invalued 'explicit-superuser' value in init message");
+        }
+
       /* Always send an init message down the new transport */
       object = cockpit_transport_build_json ("command", "init", NULL);
       json_object_set_int_member (object, "version", 1);
       json_object_set_string_member (object, "host", "localhost");
+
+      superuser = cockpit_creds_get_superuser (self->creds);
+      if (superuser && *superuser && !g_str_equal (superuser, "none"))
+        {
+          GBytes *password;
+          const guchar *password_bytes;
+          gchar *password_hex;
+          gsize password_len;
+          JsonObject *superuser_options;
+
+          password = cockpit_creds_get_password (self->creds);
+          superuser_options = json_object_new ();
+          json_object_set_string_member (superuser_options, "id", superuser);
+          if (password)
+            {
+              password_bytes = g_bytes_get_data (password, &password_len);
+              password_hex = cockpit_hex_encode (password_bytes, password_len);
+              json_object_set_string_member (superuser_options, "password", password_hex);
+              g_free (password_hex);
```

Comment:
Again, please clear the password here. This is short-lived, so presumably we don't need mlock().